pkg/vgmanager/lvm.go

```diff
@@ -23,33 +23,218 @@ import (
+}
+
 // VolumeGroup represents a volume group of linux lvm.
 type VolumeGroup struct {
+	// Name is the name of the volume group
 	Name string `json:"vg_name"`
+
+	// PVs is the list of physical volumes associated with the volume group
+	PVs []string `json:"pvs"`
 }
 
-// ListVolumeGroups lists all volume groups.
-func ListVolumeGroups(exec internal.Executor) ([]VolumeGroup, error) {
-	// volumeGroupMap type is used for marshalling the `vgs --reportformt json` command output
-	// Sample output:
-	// `{"report": [ {"vg": [{"vg_name":"fedora", "pv_count":"1", "lv_count":"3", "snap_count":"0", "vg_attr":"wz--n-", "vg_size":"<475.94g", "vg_free":"0 "}]}]`
-	volumeGroupMap := map[string][]map[string][]VolumeGroup{}
-	arg := []string{
+// Create creates a new volume group
+func (vg VolumeGroup) Create(exec internal.Executor, pvs []string) error {
+	if vg.Name == "" {
+		return fmt.Errorf("failed to create volume group. Volume group name is empty")
+	}
+
+	if len(pvs) == 0 {
+		return fmt.Errorf("failed to create volume group. Physical volume list is empty")
+	}
+
+	args := []string{vg.Name}
+	args = append(args, pvs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(createVGCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to create volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Extend extends the volume group only if new physical volumes are available
+func (vg VolumeGroup) Extend(exec internal.Executor, pvs []string) error {
+	// Extend VG only if there is any new Physical Volume
+	newPVs := listDiff(pvs, vg.PVs)
+	if len(newPVs) == 0 {
+		return ErrNoNewPhysicalVolumes
+	}
+
+	args := []string{vg.Name}
+	args = append(args, newPVs...)
+
+	_, err := exec.ExecuteCommandWithOutputAsHost(extendVGCmd, args...)
+	if err != nil {
+		return fmt.Errorf("failed to extend volume group %q. %v", vg.Name, err)
+	}
+
+	return nil
+}
+
+// Delete deletes a volume group and the physical volumes associated with it
+func (vg VolumeGroup) Delete(exec internal.Executor) error {
+	// Remove Volume Group
+	vgArgs := []string{vg.Name}
+	_, err := exec.ExecuteCommandWithOutputAsHost(vgRemoveCmd, vgArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to remove volume group %q. %v", vg.Name, err)
+	}
+
+	// Remove physical volumes
+	_, err = exec.ExecuteCommandWithOutputAsHost(pvRemoveCmd, vg.PVs...)
+	if err != nil {
+		return fmt.Errorf("failed to remove volume group %q. %v", vg.Name, err)
+	}
+	return nil
+}
+
+// GetVolumeGroup returns a volume group along with the associated physical volumes
+func GetVolumeGroup(exec internal.Executor, name string) (*VolumeGroup, error) {
+	res := new(vgsOutput)
+	args := []string{
 		"vgs", "--reportformat", "json",
 	}
-	output, err := exec.ExecuteCommandWithOutputAsHost(lvmPath, arg...)
+	if err := execute(exec, res, args...); err != nil {
+		return nil, fmt.Errorf("failed to list volume groups. %v", err)
+	}
+
+	vgFound := false
+	volumeGroup := &VolumeGroup{}
+	for _, report := range res.Report {
```

Comment:
If I try `vgs vg2 --reportformat json`, then I need to parse the output again and check for `Volume group "vg2" not found` string is present or not.   To avoid this parsing, I'm trying get all the vgs first, then a trying to find if `vg2` is one of them.

Reply:
Wouldn't the output just be an empty array?

Reply:
Looks like vgs returns exit code 5 when trying to list a non-existent VG, which just means command failed. Trying to figure out why would be more difficult than iterating over the list so the current code is fine for now.
Please add a comment that vgs can take take a vg name and that we will look into whether that can be used later.

Reply:
added comment